pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go

```diff
@@ -0,0 +1,54 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filter
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	v1 "k8s.io/client-go/listers/core/v1"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	// PilotDiscoveryLabelName is the name of the label used to indicate a namespace for discovery
+	PilotDiscoveryLabelName = "istio-discovery"
+	// PilotDiscoveryLabelValue is the value of for the label PilotDiscoveryLabelName used to indicate a namespace for discovery
+	PilotDiscoveryLabelValue = "true"
+)
+
+func DiscoveryNamespacesFilterFunc(lister v1.NamespaceLister, enableDiscoveryNamespaces bool) func(obj interface{}) bool {
```

Comment:
do you have any profiles on istiod side? to see if the filtering is taking any noticeable CPU/mem usage?

Reply:
Without the filter: 
![image](https://user-images.githubusercontent.com/3580102/103316241-e011a500-49f5-11eb-93ab-cb2a936def39.png)

With the filter:
![image](https://user-images.githubusercontent.com/3580102/103316248-eacc3a00-49f5-11eb-988d-5ef8c04f4d62.png)

There appears to be less CPU utilization (50ms -> 20ms) by the k8s watch when adding the filter

Reply:
It seems the cost is very big, what if we remove the filter when get/list

Reply:
@hzxuzhonghu Can you elaborate on where you're seeing the cost incurred by the filtering itself? Zooming in on the second profile I posted above (the one with the filter applied), it seems the additional cost of the filter consists of map accesses and string equality comparisons. But overall the filter appears to be conserving CPU usage, since there are fewer resources by processed.

![image](https://user-images.githubusercontent.com/3580102/103350619-95803f00-4a6e-11eb-8660-2bc55d18f291.png)

With respect to removing the filter from get/list, for correctness we want the filter to be applied to all reads, and there are numerous usages of `List()` and `GetByKey()` throughout the kube service registry

Reply:
I don't think the profiler can give accurate numbers at such low load

Reply:
I conducted another profile using [pilot-load](https://github.com/howardjohn/pilot-load) to generate load against pilot. With the cache-backed filter approach implemented in commit `a3d6016`, the performance impact of the filtering itself should be nominal since each invocation of the filter only requires constant time access into a map (as opposed to the previous implementation which required iteration of all namespaces).

Without filtering:

![image](https://user-images.githubusercontent.com/3580102/103557871-d074df00-4e81-11eb-8b20-75dfb0f340c3.png)

With filtering enabled and no labeled namespaces:

![image](https://user-images.githubusercontent.com/3580102/103557979-f8644280-4e81-11eb-858d-0b98a8a67f49.png)

With filtering enabled and 10 labeled namespaces:

![image](https://user-images.githubusercontent.com/3580102/103558549-d4edc780-4e82-11eb-81ae-9dbdfd0b6932.png)